### DISCUSSION TITLE
Tasmota Bluetooth 11.1 + Xiaomi LYWSD02 fw 2.0.1

### DISCUSSION THREAD

**Discussion by ceinmart** [arendst/Tasmota#20041] [Q&A] [answered]
Hi, 

I have a ESP32 with Tasmota bluetooth v11.1 installed. 
With [blerry](https://github.com/tony-fav/tasmota-blerry) installed for others two sensors with custom firmware from pvvx. 
(Using tasmota v11.1 , because unfortunately , the version 13 don't work well with blerry scripts, crash repeatable) 

I bougth a LYWSD02 sensor, tried to configure it manually using MI32Keys , where I get the bind_keys using the "token_extractor.exe" . 
But they don't recognize the Key . 
On main menu still showing : Unknown NoKey
Looking at MiHome, it became with firmware v2.0.1.

After a research, found at ESPHome they support LYWSD02 firmware v1 only , not v2.
Here in tasmota, have the same limitation too ? 
anyone know something about?? 

Is there a way to setup this sensor using tasmota bt v11 ???

**Comment by Jason2866** [comment]
Not supported.

**Comment by ceinmart** [comment]
> Not supported.

Hi @Jason2866 , please , can enlight me if this firmware v2 isn't supported or LYWSD02 at all isn't supported anymore?  
Because at tasmota documentation has no information about firmware version. 
There we understand that any version of LYWSD02 should work...

**Reply by Jason2866** [reply]
V2 firmware (which sends encrypted) is not supported
You can update Tasmota docu. As time of writing only firmware V1 did exist

**Comment by ceinmart** [comment]
Hi! 

More doub'ts about this situation: 

Tasmota have the command below to decrypt the comunication: 
`MI32Keys <mac or blealias>=<bind_key> <mac or blealias>=<key>`

Even settings they bind_key (which I get with token extractor) , still not decrypting or showing the information (I think I should see it on web admin page of my tasmota) ... 

that is because of the new fw version ?

**Reply by barbudor** [reply]
I believe no many people are using BLE outside of LYWSD03MMC with Pvvx firmware
Personnaly I would not invest time in it

**Reply by ceinmart** [reply]
Hi @barbudor  , 
Yes, I have two LYWSD03MMC with pvvx firmware running smoothly with ESP32/tasmota + blerry script sending MQQT to HA. 
However, I and many other users also have LYWSD02MMC, and everyone wants to get the data from it since they are already there.
It makes no sense to buy a D03 to put on the side only to send the data to HA.

That is why I'm not giving up to try to configure LYWSD02MMC using the tasmota.

**Reply by barbudor** [reply]
As an open source project, this may be debugged/added in Tasmota by someone who
- has an interrest (because he owns the device, and may not switch to something else)
- has the skills to work on this part of the code

I check none of the boxes.
I did contribute on the BLE driver on some Mi32Options regarding formatting of MQTT messages
D03 with PVVX are damn cheap and working so fine that I have purchased around 20 of them across 2 sites. But I gave up on all BLE buttons and door-sensor I tried because it was too hard work and prefered purchase Zigbee ones instead.

So maybe @btsimonh and @Staars who wrote major contributions to BLE, or @kabongsteve who recently contribute on some BLE buttons can be able to help.

**Comment by Staars** [comment]
I don't have this sensor either. It should work as an unknown sensor.
I can not reproduce any problems with Tasmota v13 nor with BLE remotes (core2 or 3) so I fear, I can't help here.

**Comment by ceinmart** [comment]
Hi Guys, I really appreciate all attention spent here!!! 
Thank you!!!

So, I'm trying to understand a few things that aren't clear and neither easy to find on the web. 
Tasmota has partial support for encrypted data with bind_keys for xiaomi products , right?

Happen some change on xiaomi communication when they change the firmware v1 to v2 on they sensors?
Is some cipher change or something like this which makes the MI32Keys not work with this product? 

I have no skill to work on sources or debug, but I can help if some collected data can give some direction. 
Just need some direction on how to collect this data to share.

**Comment by btsimonh** [comment]
maybe look at the pvvx github repos , and ask there?  The firmware update for the temp and hum sensors was quite supported when I was playing with mine....
There may be 'issues' there that help in TAS configuration, etc.

**Comment by ceinmart** [comment]
Hi, Just sharing ... 

I was able to setup my LYWSD02MMC on my home assistant, unfortunately, not using the Tasmota... 
I am connecting it directly to Bluetooth BLE on my H.A. box. 
I bought a Bluetooth BLE token, plugged it into my H.A. hardware (which is a box TV running armbian/debian/docker) , configured docker, the Bluetooth integration recognizes automatically the token, configured the BLE tracker integration (manually on configuration.yaml) and voila, they just get recognize the beacon from my Xiaomi device automatically, appear as new suggested device, ask for the bindkey (which I get with Xiaomi extractor) and just works!
The device broadcasts the data every 5 minutes, which is perfect for me... 

Now, if the HA has a way to make this beacon be decrypted, I think this could be "easily" implemented on Tasmota too... 

My first desire is work using tasmota + Blerry, but, now the next ESP32 I get, probably I will go to ESPHome , since they work as BLE gateway with H.A. (as far I know, they tasmota didn't), this way I get the H.A. resources to setup news BLE devices I intend to implement at home.
I really like the Tasmota, but unfortunately, they need to improve this integration and resources to be usable with H.A.... 

![image](https://github.com/arendst/Tasmota/assets/7062410/086822a1-c513-4057-beb7-85fd8746ea72)

**Comment by SteWers** [comment, marked as answer]
LYWSD02MMC is supported since version 13.3.0.3

![image](https://github.com/arendst/Tasmota/assets/42718143/b8172a29-204d-4994-8708-85fb09fca260)

**Reply by ceinmart** [reply]
You are right!
I bought a second ESP32 and installed the last version of Tastmota. Configured the bind key and works!!! 
Thanks!!

**Comment by ceinmart** [comment]
@SteWers  , 

Did you enable anything else to work with your D02?

When I configured here, they show KeyOK ... however , after waiting and waiting, no information appear until now. 
I tried enable few commands, without success. 
I also enabled this two commands: 
```
ibeacon 1 ;
BLEAddrFilter 3 ;
```

As I mentioned on previous post, this sensor already working with H.A. BLE Tracker . 
(but I need to put it to work with this ESP32 because of the distance of my H.A. hardware)

That is what I get:

<img width="284" alt="image" src="https://github.com/arendst/Tasmota/assets/7062410/b0cb3a95-aa6a-4fea-ac09-dc45b1eac068">

At the console , this is what appear  : 
```

19:45:00.678 MQT: tele/esp32_ble_tracker2/SENSOR = {"Time":"2024-01-13T19:45:00","IBEACON":{"MAC":"A4C1383CF1CE","NAME":"SensMi_Sala","RSSI":-66,"STATE":"ON","PERSEC":0}}
19:45:12.830 M32 A4C1383CF1CE: Payload encrypted but no key
19:45:12.833 M32: MI packet with MAC addr mismatch - is this mesh?

19:45:41.956 MQT: tele/esp32_ble_tracker2/SENSOR = {"Time":"2024-01-13T19:45:41","Unknown3cf1ce":{"alias":"SensMi_Sala","mac":"a4c1383cf1ce","RSSI":-76}}

```

**Reply by SteWers** [reply]
There is nothing special to configure, but you should use the lastest [dev version](https://ota.tasmota.com/tasmota32/tasmota32-bluetooth.bin). Maybe you are on the release version:
![image](https://github.com/arendst/Tasmota/assets/42718143/c1f67fa3-1f22-4d2c-ab12-ac121774c8bc)
It should look like this:
![image](https://github.com/arendst/Tasmota/assets/42718143/2c323ddd-dcd1-4d9f-be0f-31798f42dcc8)

**Reply by ceinmart** [reply]
Thank you! That is it!!! 
Now really works!
Next challenge, understand how to make the device and data shows on my MQTT ... without use [blerry ](https://github.com/tony-fav/tasmota-blerry) scripts
<img width="314" alt="image" src="https://github.com/arendst/Tasmota/assets/7062410/300144ba-aba2-4127-b63b-c2836df092bb">

**Reply by SteWers** [reply]
Nice to hear. 😊
The MQTT message (tele/SENSOR) is generated automatically. Of course you have to configure MQTT in Tasmota first. Here is a sample output:
```
20:40:28.969 MQT: tele/ESP32BLE/SENSOR = {"Time":"2024-01-14T20:40:28","LYWSD02MMCc549a3":{"alias":"Uhr","mac":"a4c138c549a3","Temperature":16.4,"Humidity":44.0,"DewPoint":4.1,"Battery":74,"RSSI":-93},"TempUnit":"C"}
```

**Reply by ceinmart** [reply]
@SteWers 

Have you setup blerry scripts on your ESP32 ? 
Or you add your xiaomi devices on HA using other method?

**Reply by SteWers** [reply]
I don't use HA. I work with openHAB. Maybe automatically discovery is a good thing, but I prefer to add devices per hand. Hopefully I will stay a little bit more smart than my home. Since the number of devices does not increase by several hundreds per day, this is the best way for me. 😉

**Reply by Lassska** [reply]
Hi,
see my answer below

**Comment by ceinmart** [comment]
Yes , my MQTT is configured and receiving theses messages... 
However, they don't create the device... 
![image](https://github.com/arendst/Tasmota/assets/7062410/bcf69d6e-28e2-47de-b220-7b485d391000)


<img width="812" alt="image" src="https://github.com/arendst/Tasmota/assets/7062410/c7234683-5105-417d-985e-8b143b381428">


Looking at the logs, I found nothing... 
I ran my H.A. on Docker and MQTT is a Docker container... 
I looked at the MQTT log too, also found nothing relevant, only messages from the mqtt clients (my esp32 devices)

Trying to figure out , what is happen. 
I have 3 BLE tracker on my H.A.  , one dongle on H.A. Hardware + 2 ESP32 with tasmota.
This sensor, for test I added it using the H.A. dongle , where they recognize and suggest to add automatically the integration "Xiaome BLE".
I don't know if the H.A. identifies the MAC and doesn't add the device to not duplicate (!?), so, I deleted it, restarted my H.A. ... 
Still not appearing on my MQTT... 
They receive the message but do not create the device...

**Reply by sfromis** [reply]
HA automatic device discovery of devices served via Tasmota depends on extra discovery payloads from Tasmota right after boot, where no BLE sensors are available yet.

**Reply by ceinmart** [reply]
Hi @sfromis  , 

Sorry , I really don't understand your explanation. 
If your explanation is true, how the BLE devices are discovered after all? 
All these setup of tasmota + mqtt appears to be very confusing for me...  looks simple, like plug-in-play, however appears not to be.

Is it because of that, we needed to use blerry scripts?  (which I use with other ESP32/tasmota 11.1 )

**Reply by barbudor** [reply]
Yes
Tasmota Discovery message announce to HA only the native sensors (those physically attached to the ESP)
For BLE sensors you need to use Blerry
For Zigbee sensors there isn't an equivalent of Blerry yet
Someone has to do it

**Reply by btsimonh** [reply]
Another options is to regularly send a discovery message from something like Node-Red.  I.e. have node-red add all the sensors you expect to be there.  And if your Tas's use the same MQTT topic for the same device, then it does not matter which TAS it comes from.
(I hope HA has improved.  This was HELL 3 years ago when I tried it.  I thnk HA has come on quite a bit.).

**Reply by ceinmart** [reply]
Thanks @barbudor !
I get a little deep here on MQTT broker configuration and see to discovery , they need to receive into the topic /homeassistant/status
And I really don't see it here... 
Will install blerry here and hope they work... with my last ESP32 using the last Tasmota release, they just reboot every 10sec.  Need to downgrade it to version 11.1 to make work (recommendation from other user on other topic).
However, I'm using sensors with customized firmware there,  which is simple...

**Reply by barbudor** [reply]
That's the old way which is no more supported

Now Tasmota Discovery publishes on tasmota/
It requires that you install the Tasmota Integration plug-in also known as HaTasmota

**Reply by ceinmart** [reply]
I have the Tasmota integration, my esp32 appears there immediately ... but that is it... no more resources, devices... 
However, as you explanation, I understand that doesn't solve the BLE issue , still need the blerry scripts. 

Just so everyone understands my insistence here, try to avoid installing *blerry* scripts...


<img width="473" alt="image" src="https://github.com/arendst/Tasmota/assets/7062410/2caa8dca-6733-4cf2-84cd-0f38cccb8f7e">

<img width="798" alt="image" src="https://github.com/arendst/Tasmota/assets/7062410/96bbbad7-1dd3-48a4-ba08-f7d5c4bcf0f6">

**Comment by bam80** [comment]
> Using tasmota v11.1 , because unfortunately , the version 13 don't work well with blerry scripts, crash repeatable

Is it this crash or some other?
https://github.com/tony-fav/tasmota-blerry/issues/74

To others:
what is the problem here, is it on Tasmota or BLErry side?

**Reply by sfromis** [reply]
That thread does not have sufficient info to guess about where the issue is. In general, flaws in Berry code should not lead directly to such a crash, but when things gets beyond simple, it would be perfectly possible for Berry code to create a situation which core Tasmota code could not handle.

FWIW (which may not be much), I saw no crash when simply doing the suggested install of the project. As I did no setup of devices, my situation may be too simple to be a useful test.

In general, such issues should be debugged, including giving specific info about how a failure can be recreated on other devices.

**Reply by ceinmart** [reply]
Hi @bam80  , 
No, isn't the same situation. You can see the panic [log here](https://github.com/tony-fav/tasmota-blerry/discussions/28#discussioncomment-7466577) .
However, with this last version of Tastmota (dev), I also get other problem using blerry which my Tasmota begin to reboot repeatable, without stop. This time, I was able to workaround the problem and stop the reboots.  On this ESP I've configured BLEAlias for all my devices and for some reason which I don't know why, blerry scripts give the error below repeatable until freeze and reboot (this time I don't connect to usb/serial to get the panic stack trace, have only the log send to my rsyslog).
My workaround was not to set the BLEAlias to A4C13891BFB7 , then they work , not 100% , but work.
Now I have other situation, blerry isn't send all info to H.A . But I think [this is a blerry problem](https://github.com/tony-fav/tasmota-blerry/discussions/22#discussioncomment-8149025).


```
0045E2B25B","0xfe95":"585842255FCEF13C38C1A40B1F07A0BB430045E2B25B"}}
Jan 16 16:51:31 esp32-ble-tracker2 ESP-BLY: Xiaomi: Cannot support encrypted data.
Jan 16 16:51:32 esp32-ble-tracker2 ESP-RSL: BLE = {"DetailsBLE":{"mac":"A4C13891BFB7","a":"SensMi_B7_SMaster","RSSI":-89,"p":"02010612161A18B7BF9138C1A4DF0B7418DA0B64FD04","0x181a":"B7BF9138C1A4DF0B7418DA0B64FD04"}}
Jan 16 16:51:32 esp32-ble-tracker2 ESP-BLY: tried to handle mac = A4C13891BFB7 with alias = SensMi_B7_SMaster
Jan 16 16:51:32 esp32-ble-tracker2 ESP-BRY: Exception> 'key_error' - A4C13891BFB7
Jan 16 16:51:32 esp32-ble-tracker2 ESP-stack traceback:
Jan 16 16:51:32 esp32-ble-tracker2 ESP-#011<unknown source>: in function `handle_BLE_packet`
Jan 16 16:51:32 esp32-ble-tracker2 ESP-#011<unknown source>: in function `<lambda>`
Jan 16 16:51:32 esp32-ble-tracker2 ESP-#011<unknown source>: in function `try_rule`
Jan 16 16:51:32 esp32-ble-tracker2 ESP-#011<unknown source>: in function `exec_rules`
Jan 16 16:51:32 esp32-ble-tracker2 ESP-#011<unknown source>: in function `event`
Jan 16 16:51:35 esp32-ble-tracker2 ESP-RSL: BLE = {"DetailsBLE":{"mac":"A4C1383CF1CE","a":"SensMi_Sala","RSSI":-71,"p":"020106161695FE48584225628E7D2D5B6C4E8EBB4300B1F3DCFF","0xfe95":"48584225628E7D2D5B6C4E8EBB4300B1F3DCFF"}}

```

**Reply by btsimonh** [reply]
Bear in mind any use of BLE and Wifi is (was?) subject to random crashes at the radio sharing level.
If you MUST use them together, minimise wifi usage as much as you can... and accept that it may reboot occasionally.

**Comment by Lassska** [comment]
![Screenshot_20250425_144909_Chrome](https://github.com/user-attachments/assets/69b5c44c-7751-4427-b1f7-63eb3e3e2129)
Hi all,
I was looking for a solution that was as simple as possible but without depending on the xiaomi servers for LYWSD02MMC Having a temperature sensor on an ESP32-Tasmota in the kitchen, I activated BLE and I noticed that at certain periods of no less than 5 minutes the clock sensor transmits temperature and humidity without an active key, so I collected these transmissions via mqtt (no need blerry script!) and I found that the resulting BLE temperature sensor (red or yellow graphic) does not have the same fluctuations as the same sensor connected via Xiaomi BLE integration with bind-key (blue) but I do not know for what reason it has readings higher by 0.1 °C. So I resorted to a correction with -0.1 in yaml

![Screenshot_20250425_103933_Chrome](https://github.com/user-attachments/assets/9d8730da-42f7-4699-bfbf-338afac1f0b6)
![Screenshot_20250425_104602_Gallery](https://github.com/user-attachments/assets/c999eb3f-677e-48ad-944c-4d527d63d0d1)
![Screenshot_20250425_112406_Chrome](https://github.com/user-attachments/assets/05e1b64c-1a4a-4e73-9ebb-5b975db926b0)
![Screenshot_20250425_132709_Chrome](https://github.com/user-attachments/assets/5c18f5da-2c5b-4aad-8baf-243387cdae30)
